pkg/sdkserver/sdkserver_test.go

```diff
@@ -678,176 +685,185 @@ func TestSDKServerUpdateEventHandler(t *testing.T) {
 	})
 
-	t.Run("timeout", func(t *testing.T) {
-		m, sc, stop := setup(t, defaultGs.DeepCopy())
-		defer close(stop)
+	m.AgonesClient.AddReactor("update", "gameservers", func(action k8stesting.Action) (bool, runtime.Object, error) {
+		ua := action.(k8stesting.UpdateAction)
+		gs := ua.GetObject().(*v1alpha1.GameServer)
 
-		now := time.Now()
-		fc := clock.NewFakeClock(now)
-		sc.clock = fc
+		state <- gs.Status
 
-		done := make(chan struct{}, 10)
+		return true, gs, nil
+	})
 
-		m.AgonesClient.AddReactor("update", "gameservers", func(action k8stesting.Action) (bool, runtime.Object, error) {
-			defer func() {
-				done <- struct{}{}
-			}()
+	sc, err := defaultSidecar(m)
 
-			// move past timeout
-			fc.Step(40 * time.Second)
-			return true, nil, k8serrors.NewConflict(schema.ParseGroupResource(""), "gameservers", errors.New("contention"))
-		})
+	assert.NoError(t, err)
+	stop := make(chan struct{})
+	sc.informerFactory.Start(stop)
+	assert.True(t, cache.WaitForCacheSync(stop, sc.gameServerSynced))
 
-		_, err := sc.Allocate(context.Background(), &sdk.Empty{})
-		assert.EqualError(t, err, "Allocation request timed out")
+	wg := sync.WaitGroup{}
+	wg.Add(1)
 
+	go func() {
+		err = sc.Run(stop)
+		assert.Nil(t, err)
+		wg.Done()
+	}()
+
+	assertStateChange := func(expected v1alpha1.GameServerState, timeout time.Duration, additional func(status v1alpha1.GameServerStatus)) {
 		select {
-		case <-done:
-		case <-time.After(5 * time.Second):
-			assert.Fail(t, "have contention")
+		case current := <-state:
+			assert.Equal(t, expected, current.State)
+			additional(current)
+		case <-time.After(timeout):
+			assert.Fail(t, "should have gone to Reserved by now")
 		}
-	})
+	}
+	noAdditional := func(status v1alpha1.GameServerStatus) {}
 
-	t.Run("unhealthy gameserver", func(t *testing.T) {
-		gs := defaultGs.DeepCopy()
-		gs.Status.State = v1alpha1.GameServerStateUnhealthy
-		_, sc, stop := setup(t, gs.DeepCopy())
-		defer close(stop)
+	// 3 seconds
+	_, err = sc.Reserve(context.Background(), &sdk.Duration{Seconds: 3})
+	assert.NoError(t, err)
 
-		_, err := sc.Allocate(context.Background(), &sdk.Empty{})
-		assert.EqualError(t, err, "cannot Allocate an Unhealthy GameServer")
+	assertStateChange(v1alpha1.GameServerStateReserved, 2*time.Second, func(status v1alpha1.GameServerStatus) {
+		assert.Equal(t, time.Now().Add(3*time.Second).Round(time.Second), status.ReservedUntil.Time.Round(time.Second))
 	})
 
-	t.Run("Shutdown gameserver", func(t *testing.T) {
-		gs := defaultGs.DeepCopy()
-		gs.Status.State = v1alpha1.GameServerStateShutdown
-		_, sc, stop := setup(t, gs.DeepCopy())
-		defer close(stop)
-
-		_, err := sc.Allocate(context.Background(), &sdk.Empty{})
-		assert.EqualError(t, err, "cannot Allocate a Shutdown GameServer")
+	// wait for the game server to go back to being Ready
+	assertStateChange(v1alpha1.GameServerStateRequestReady, 4*time.Second, func(status v1alpha1.GameServerStatus) {
```

Comment:
Unit tests are a lot more reliable if you don't rely on the system clock and hard coding times to wait for. Can you cause this to happen immediately if you mock the time and just advance the clock by 3 seconds immediately?

Reply:
Interesting idea I always had this problem,.

Reply:
I'm ok with this for now, but as you said in slack the presubmit tests are getting longer and longer (30+ minutes) and need some optimization. Maybe this isn't the most important place to start, but optimizing unit tests should also be on that list.

As for changing the clock interface... there's a cost of taking on new dependencies (check out https://research.swtch.com/deps if you haven't seen it) and I'm not sure that it's worth it in this case. It might be worth asking apimachinery to update their interface, but even if they did it today we wouldn't see the benefit for quite a while.

Reply:
Yeah, I had both of those thoughts as well. Okay cool - so agreed on leaving this as is. Maybe this is something we can add to the apimachinery mock down the road.